uniffi_macros/src/lib.rs

```diff
@@ -112,3 +119,100 @@ pub fn include_scaffolding(component_name: proc_macro::TokenStream) -> proc_macr
+/// available both to other Rust consumers (via the written Rust code), and to bindings from
+/// foreign languages (via the generated FFI).
+///
+/// Ideally, this macro would work as an *inner* attribute on a module, so that users
+/// could write this at the top-level of their Rust code rather than using an inline
+/// submodule:
+///
+/// ```rs
+///
+/// #![uniffi::declare_interface]
+///
+/// // The Rust code to implement the component just follows the macro.
+///
+/// pub struct Example {
+///     pub fn hello(&self) -> String {
+///         String::new("world")
+///     }
+/// }
+/// ```
+///
+/// Unfortunately, inner attributes at the top level of a crate are still unstable, and
+/// even when enabled are quite fiddly to work with. But maybe one day..!
+///
+#[proc_macro_attribute]
+pub fn declare_interface(
+    _attrs: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    // We expect to be applied to an inline module declaration.
+    let mut module = syn::parse_macro_input!(item as syn::ItemMod);
+    // The contents of the module must declare a UniFFI-compatible interface.
+    // TODO: In theory, we should be able to handle errors from parsing out the ComponentInterface
+    // and report them to the user directly on their provided Rust code, in the same way that
+    // `rustc` does. That probably requires some cleaning up of our parsing logic though.
+    let ci = ComponentInterface::from_rust_module(&module)
+        .expect("Failed to parse a ComponentInterface from the provided Rust code");
+    // Generate the Rust scaffolding.
+    // For macro purposes it would be better to generate this directly via `quote!`
+    // rather than rendering to a string and re-parsing, but this'll do for now.
+    // TODO: again, errors should be reported as red squiggles on the user's Rust code.
+    let scaffolding = RustScaffolding::new(&ci)
+        .render()
+        .expect("Failed to generated Rust scaffolding");
+    let scaffolding: syn::File =
+        syn::parse_str(scaffolding.as_str()).expect("Failed to parse generated Rust scaffolding");
+    // Append the generated scaffolding to the inline module.
+    // Unwrapping is safe because, if the module didn't have content, we wouldn't have parsed it.
+    module.content.as_mut().unwrap().1.extend(scaffolding.items);
+    // Import the public items form the module so they're available at the top level,
+    // where you might normally expect them when writing Rust by hand.
+    let namespace = &module.ident;
+    let imports = ci
+        .iter_member_names()
+        .iter()
+        .map(|name| {
+            let name = syn::Ident::new(name, module.span());
+            quote! {
+                pub use #namespace::#name;
+            }
+        })
+        .collect::<Vec<_>>();
+    proc_macro::TokenStream::from(quote! {
+        #module
+        #(#imports)*
```

Comment:
For my own understanding, any reason we're exposing those functions at the top level? Is it for other rust consumers, in which case wouldn't it be better to just keep them inside the module where consumers can see them in the code file?

Reply:
In my head, I really want the UniFFI component interface to *also* be the public interface exposed by the crate, which is why I'm lifting them to the top-level here. But I agree that's probably weird/confusing for Rust consumers since the names magically appear in a scope other than where they're defined.